libraries/iio/iio.c

```diff
@@ -311,132 +273,55 @@ static ssize_t iio_read_all_attr(void *device, char *buf, size_t len,
+		attributes[i]->store(params->dev_instance, (params->buf + j),
+				     attr_length, params->ch_info);
 		j += attr_length;
 		if (j & 0x3)
 			j = ((j >> 2) + 1) << 2;
 		i++;
 	}
 
-	return len;
-}
-
-/**
- * @brief Read/write channel attribute.
- * @param el_info - Structure describing element to be written.
- * @param buf - Read/write value.
- * @param len - Length of data in "buf" parameter.
- * @param channel - Structure describing channel attributes.
- * @param is_write -If it has value "1", writes attribute, otherwise reads
- * 		attribute.
- * @return Length of chars written/read or negative value in case of error.
- */
-static ssize_t iio_rd_wr_channel_attribute(struct element_info *el_info,
-		char *buf, size_t len,
-		struct iio_channel *channel, bool is_write)
-{
-	int16_t attribute_id;
-	struct iio_interface *iface = iio_get_interface(el_info->device_name);
-	const struct iio_ch_info channel_info = {
-		iio_get_channel_number(el_info->channel_name),
-		el_info->ch_out
-	};
-
-	if (!strcmp(el_info->attribute_name, "")) {
-		/* read / write all channel attributes */
-		if (is_write)
-			return iio_write_all_attr(iface->dev_instance, buf, len, &channel_info,
-						  channel->attributes);
-		else
-			return iio_read_all_attr(iface->dev_instance, buf, len, &channel_info,
-						 channel->attributes);
-	} else {
-		/* read / write single channel attribute, if attribute found */
-		attribute_id = iio_get_attribute_id(el_info->attribute_name,
-						    channel->attributes);
-		if (attribute_id >= 0) {
-			if (is_write)
-				return channel->attributes[attribute_id]->store(iface->dev_instance,
-						(char*)buf, len, &channel_info);
-			else
-				return channel->attributes[attribute_id]->show(iface->dev_instance, (char*)buf,
-						len, &channel_info);
-		}
-	}
-
-	return -ENOENT;
+	return params->len;
 }
 
 /**
  * @brief Read/write attribute.
- * @param el_info - Structure describing element to be written.
- * @param buf - Read/write value.
- * @param len - Length of data in "buf" parameter.
- * @param iio_device - Physical instance of a device.
+ * @param params - Structure describing parameters for store and show functions
+ * @param attributes - Array of attributes.
+ * @param attr_name - Attribute name to be modified
  * @param is_write -If it has value "1", writes attribute, otherwise reads
  * 		attribute.
  * @return Length of chars written/read or negative value in case of error.
  */
-static ssize_t iio_rd_wr_attribute(struct element_info *el_info, char *buf,
-				   size_t len,
-				   struct iio_device *iio_device, bool is_write)
-{
-	int16_t channel_id;
-	int16_t attribute_id;
-	struct iio_interface *iface;
-
-	if (!iio_device)
+static ssize_t iio_rd_wr_attribute(struct attr_fun_params *params,
+				   struct iio_attribute **attributes,
+				   char *attr_name,
+				   bool is_write)
+{
+	struct iio_attribute *attr;
+
+	/* Search attribute */
+	for (attr = attributes[0]; attr; attr++)
```

Comment:
Are we sure that the `attributes` table will always end in NULL pointer?

Reply:
Yes, in linux is this way so we will use the same pattern.